Update multiple-focus-metavariables.md (#737)
returntocorp/semgrep-docs: docs/experiments/multiple-focus-metavariables.md

description: "With this rule, Semgrep matches all pieces of code captured by the focus metavariables."
---

> # Using multiple focus metavariables with set union semantics

Author: ```suggestion
# Using multiple focus metavariables using set union semantics
```

# Using multiple focus metavariables

Semgrep matches all pieces of code captured by focus metavariables when you specify them in a rule. Include the metavariables that you want to focus on by specifying each in a YAML list. 

:::info
This feature is using `focus-metavariable`, see [`focus-metavariable`](/writing-rules/rule-syntax/#focus-metavariable) documentation for more information.  
:::

> 

Author: ```suggestion
There are two ways in which you can include multiple focus metavariables:

```

> # Using multiple focus metavariables

Reviewer: Perhaps I'm nitpicking, but I would call this experiment "`focus-metavariable` with union semantics` or something like that.

Reviewer 2: I think using the word "semantics" instead of "syntax" makes more sense here

> Semgrep matches all pieces of code captured by focus metavariables when you specify them in a rule. Include the metavariables that you want to focus on by specifying each in a YAML list or by including more focus-metavariables. 

Reviewer: ```suggestion
```

:::info
This feature is using `focus-metavariable`, see [`focus-metavariable`](/writing-rules/rule-syntax/#focus-metavariable) documentation for more information.  
:::

> There are two methods of focusing Semgrep on multiple metavariables. Each of them can lead to different results:

Reviewer: ```suggestion
```

Author: Why not?

Reviewer 2: Oh, I guess the "and"/"advanced" method doesn't actually focus in on multiple, it just focuses in on whichever has a match with a smaller range if their ranges overlap.

Author: @enncoded this text is no longer in the document, can we comment on the current state?

## Set union

> Include `focus-metavariable` with more metavariables directly under the `patterns` key to match these metavariables regardless of their position in code.

Author: ```suggestion
Include `focus-metavariable` with more metavariables directly under the `patterns` key to match these metavariables regardless of their position in code.

For example, there is a pattern that is binding several metavariables. You want to produce matches focused on two or more of these metavariables. If you specify a list of metavariables under `focus-metavariable`, each focused metavariable matches code independently of the others.
```

> Include `focus-metavariable` with more metavariables directly under the `patterns` key to match these metavariables regardless of their position in code.

Reviewer 2: I think "regardless of their position in code" isn't actually the main thing to emphasize here, it's moreso a capability that you get by using this feature.

> Include `focus-metavariable` with more metavariables directly under the `patterns` key to match these metavariables regardless of their position in code.

Author: ```suggestion
For example, there is a pattern that is binding several metavariables. You want to produce matches focused on two or more of these metavariables. If you specify a list of metavariables under `focus-metavariable`, each focused metavariable matches the code independently of the others.
```

```yaml
    patterns:
      - pattern: foo($X, ..., $Y)
      - focus-metavariable: 
        - $X
        - $Y
```

> This syntax enables Semgrep to match these metavariables regardless of their position in code. See the following example:

Reviewer: ```suggestion
The example above is equivalent to:

    pattern-either:
    - patterns:
       - pattern: foo($X, ..., $Y)
       - focus-metavariable: $X
    - patterns:
      - pattern: foo($X, ..., $Y)
      - focus-metavariable: $Y
```

Author: Implemented!

<iframe src="https://semgrep.dev/embed/editor?snippet=D602" border="0" frameBorder="0" width="100%" height="432"></iframe>

:::tip
Among many use cases, the **set union** syntax allows you to simplify taint analysis rule writing. For example, see the following rule:
<iframe src="https://semgrep.dev/embed/editor?snippet=w6Qx" border="0" frameBorder="0" width="100%" height="432"></iframe>
:::

> ## Set intersection

Reviewer: I'm not sure this needs to be here since it's not an experiment, it's a GA feature.

Author: @IagoAbal We are trying to distinguish this in a single article. Moving these sections around can look quite chaotic.
